understanding-tidb: add introduction for statistics (#70)
pingcap/tidb-dev-guide: src/understand-tidb/table-statistics.md

### Histogram

Histogram splits the data into a number of buckets and maintains some information to describe a bucket, such as the number of records in the bucket. It is widely used in many RDBMS to estimate a range.

Among two commonly used histogram strategies, equal-depth and equal-width, we choose the equal-depth histogram for it has a better guarantee of worst case error rate compared to that of the equal-width histogram. You can refer to the paper [Accurate estimation of the number of tuples satisfying a condition](https://dl.acm.org/citation.cfm?id=602294) for more information. With the equal-depth histogram, the number of values in each bucket is to be as equal as possible. For example, to split the given record set of `1.6, 1.9, 1.9, 2.0, 2.4, 2.6, 2.7, 2.7, 2.8, 2.9, 3.4, 3.5` into 4 buckets, the final buckets would look like `[1.6, 1.9], [2.0, 2.6], [2.7, 2.8], [2.9, 3.5]`.	Thus the depth, (a.k.a. the number of records) of each bucket is 3, as shown in the following graph.

![equal-depth histogram](/src/img/stats-histogram.png)

> ###Count-Min Sketch 

Reviewer: ```suggestion
### Count-Min Sketch 
```

The Count-Min Sketch (CM sketch) is a data structure used for query cardinality estimation for the equal predicate, or join, etc., and provides strong accuracy guarantees. Since its introduction in 2003 in the paper [An improved data stream summary: The count-min sketch and its applications](http://dimacs.rutgers.edu/~graham/pubs/papers/cm-full.pdf), it has gained widespread use given its simplicity of construction and use.

CM sketch maintains an array of `d*w` counts, and for each value, maps it to a column in each row using `d` separate hash functions and modifies the count value at those `d` positions. This is shown in the following figure.

![count-min sketch](/src/img/stats-cmsketch.png)

This way, when querying how many times a value appears, the d hash functions are still used to find the position mapped to in each row, and the minimum of these d values is used as the estimate. 

**Please note that CM sketch is not used as default statistics since version 5.1 given the increasing concerns on estimation bias under the scenarios with large distinct values of a column.**

> 
> 

Reviewer: ```suggestion

```

# TiDB Statistics

In the TiDB's SQL optimization phase, we will have the cases like which index of one table is most suitable one, whether a join operator should be a hash join, a index nested loop join or other join algorithm. Thus we need the statistics of the table to know its data distribution, guiding the TiDB choose the optimial execution plan.

In this article, we would first introduce the data structure that database could use then have a easy guide of TiDB's statistics.

## Statistics introduction

> There're many data structures to show the distribution of the data. Here we introduce three kind of them: histogram, count-min sketch and top-n value(most frequnt values).

Reviewer: ```suggestion
There're many data structures to show the distribution of the data. Here we introduce three kinds of them: histogram, count-min sketch and top-n value(most frequnt values).
```

Are they all kinds of statistics?

> ### Histogram

Reviewer: Where is the related code about histogram?

#### Introduction

Histogram splits the data into many buckets and uses some simple things to describing the bucket, such as how many records in in this bucket. It's widely used in many RDBMS to do the range estimation. We have two different type of histogram depending on the bucketing strategy: equal-depth histogram and equal-width histogram.

> We choose the equal-depth histogram according to the paper [Accurate estimation of the number of tuples satisfying a condition](https://dl.acm.org/citation.cfm?id=602294). The equal-depth histogram has a better guarantee of the error rate compared in the worst cases, compared with the equal-width histogram. The so-called equal-depth histogram means that the number of values falling into each bucket is as equal as possible. For example, we want to split the given records set `1.6, 1.9, 1.9, 2.0, 2.4, 2.6, 2.7, 2.7, 2.8, 2.9, 3.4, 3.5` into 4 buckets. Then we would get the final result as `[1.6, 1.9], [2.0, 2.6], [2.7, 2.8], [2.9, 3.5]`, the depth of each bucket is 3, i.e. the number of records in each bucket is 3. The graph is shown as below.

Reviewer: It's better to describe what "equal-width" means here, though it is straight forward.

Author: I think the deeper information can be found in the paper. Also, if we introduce the equal-width case, we need to tell the reader why we choose the equal-depth one.

Reviewer: So you can write "the deeper information can be found in the paper, in conclusion, we choose the equal-depth histogram..." or so.

Reviewer: Because a reader to here will think, you wrote "We have two different type of histogram depending on the bucketing strategy: equal-depth histogram and equal-width histogram." above, and talk one of them, why?

You comment here is better to be present on the content instead of a PR comment.

### Count-Min Sketch(Legacy in TiDB)

The Count-Min Sketch is a data structure that can handle equivalence queries, join size estimation, etc., and provides strong accuracy guarantees. Since its introduction in 2003 in the paper [An improved data stream summary: The count-min sketch and its applications](http://dimacs.rutgers.edu/~graham/pubs/papers/cm-full.pdf), it has gained widespread use due to its simplicity of creation and use.

Count-Min Sketch maintains an array of d*w counts, and for each value, maps it to a column in each row using d separate hash functions, and modifies the count value at those d positions. This is shown in the following figure.

![count-min sketch](/src/img/stats-cmsketch.png)

This way, when querying how many times a value appears, the d hash functions are still used to find the position mapped to in each row, and the minimum of these d values is used as the estimate.

> ### Top-N value(Most Frequent Value)

Reviewer: Where is the related code about top-n value?

Therefore, one of the most important tasks of Selectivity is to divide all the query conditions into as few groups as possible, so that the conditions in each group can be estimated using the statistics on a column or index, so that we can make as few independence assumptions as possible.

We use a simple greedy algorithm do the grouping. We always choose the index/column which cover the most remaining filters and remove these filters, go into next round until all filters are covered. The last step is to do the estimation using the statistics on each column and each index as mentioned before, and combine them with the independence assumption as the final result.

## Summary

> The collection and maintenance of statistics is the core function of the database. And for the cost-based query optimizer, the accuracy of statistics directly affects the query performance. In the distributed database, collecting statistics is not much different from standalone, but maintaining statistics has more challenges, such as how to maintain accurate and timely statistics in the case of multi-node updates.

Reviewer: ```suggestion
The collection and maintenance of statistics is one of the core functions of the database. And for the cost-based query optimizer, the accuracy of statistics directly affects the query performance. In the distributed database, collecting statistics is not much different from standalone, but maintaining statistics has more challenges, such as how to maintain accurate and timely statistics in the case of multi-node updates.
```

For dynamic updating of histograms, the industry generally has two approaches.

- For each addition or deletion, go to update the corresponding bucket depth. Splitting a bucket when its depth is too high is generally done by dividing the width of the bucket equally, although this makes it difficult to accurately determine the splitting point and causes errors.

> - Using the actual number obtained from the query to adjust the histogram with feedback assumes that the error contributed by all buckets is uniform, and uses the continuous value assumption to adjust all the buckets involved. However the assumption of uniformity of errors often causes problems, such as when a newly inserted value is larger than the maximum value of the histogram, it will spread the error caused by the newly inserted value to the histogram, thus causing errors.

Reviewer: why not move these paragraphs to the section of histogram

Author: If so, what should be placed here?

![equal-depth histogram](/src/img/stats-histogram.png)

#### Constrcution

The construction will vary depending on whether we know the size of the complete dataset.

If we know the size, we can dicide the depth of the bucket. So we just iterating each value `v`:

- If `v` equals to the previous one, we just place it to the same bucket no matter the bucket is full or not. This can make sure that the same value is always in the same bucket.
- If not, we check whether the bucket is full or not. If not full, put it into the current bucket. Otherwise we put it to a new bucket.

If we don't know the size, we construct the histogram in the following way. We initialize the bucket depth to 1 for each bucket. And just inserting the data like before. Once the we meet the case that one bucket exceeds the needed depth, we double the depth of the bucket and combine two adjacent buckets into one bucket.

> ### Count-Min Sketch(Legacy in TiDB)

Reviewer 2: Why it becomes legacy?

Reviewer: ... and where is the related code about count-min sketch?

Author: @xxchan The explanation is written at the place where we introduce the top-n.

Reviewer: @winoros so you can mention here briefly that it is superseded by top-n?
